src/cpp/src/block_manager.hpp

```diff
@@ -75,56 +78,81 @@ class KVCacheBlock {
+        return lhs.second[0]->get_timestamp() < rhs.second[0]->get_timestamp();
     }
 
-    KVCacheBlock::Ptr get_block(size_t hash) {
+    std::vector<KVCacheBlock::Ptr> get_block(size_t hash) {
         if (blocks.find(hash)== blocks.end())
         {
-            return nullptr;
+            return {};
+        }
+        std::vector<KVCacheBlock::Ptr> blocks_for_all_layers = blocks[hash];
+        for (auto& block_ptr : blocks_for_all_layers) {
+            block_ptr->set_timestamp(std::chrono::system_clock::now());
+            block_ptr->increment();
         }
-        KVCacheBlock::Ptr block = blocks[hash];
-        block->set_timestamp(std::chrono::system_clock::now());
-        block->increment();
         blocks.erase(hash);
-        return block;
+        return blocks_for_all_layers;
     }
 
-    KVCacheBlock::Ptr get_lru_block() {
-        if (!blocks.size()) {
-            return nullptr;
+    std::vector<KVCacheBlock::Ptr> get_lru_block() {
+        if (blocks.empty()) {
+            return {};
+        }
+        auto hash_and_blocks_for_all_layers = std::min_element(std::begin(blocks), std::end(blocks), block_is_less);
+        auto blocks_for_all_layers = hash_and_blocks_for_all_layers->second;
+        auto timestamp = std::chrono::system_clock::now();
+        for (auto& block_ptr : blocks_for_all_layers) {
+            block_ptr->set_timestamp(timestamp);
+            block_ptr->increment();
         }
-        auto hash_block = std::min_element(std::begin(blocks), std::end(blocks), block_is_less);
-        auto block = hash_block->second;
-        block->set_timestamp(std::chrono::system_clock::now());
-        block->increment();
-        blocks.erase(hash_block->first);
-        return block;
+        blocks.erase(hash_and_blocks_for_all_layers->first);
+        return blocks_for_all_layers;
     }
 
     size_t num_blocks() const {
         return blocks.size();
     }
 };
 
+class CacheStateDumper;
 
 class BlockAllocator {
-    std::list<KVCacheBlock::Ptr> m_free_blocks;
-    ov::genai::Evictor m_evictor;
+    std::vector<std::list<KVCacheBlock::Ptr>> m_free_blocks;
     int m_total_num_blocks;
+    friend class CacheStateDumper;
+    size_t m_num_layers;
     bool m_enable_prefix_caching;
+    ov::genai::Evictor m_evictor;
 public:
-    BlockAllocator(int num_blocks, bool enable_prefix_caching) :
-        m_total_num_blocks(num_blocks), m_enable_prefix_caching(enable_prefix_caching) {
-        for (int block_id = 0; block_id < m_total_num_blocks; ++block_id) {
-            m_free_blocks.push_back(std::make_shared<KVCacheBlock>(block_id));
+    BlockAllocator(int num_blocks, bool enable_prefix_caching, size_t num_layers = 0) :
+        m_total_num_blocks(num_blocks), m_num_layers(num_layers), m_enable_prefix_caching(enable_prefix_caching), m_evictor(num_layers) {
+        if (m_num_layers == 0) {
```

Comment:
maybe default value for n_num_layers would be the right choice?

Reply:
Done